locust/test/test_runners.py

```diff
@@ -734,11 +743,15 @@ def test_spawn_locusts_in_stepload_mode(self):
             for i in range(5):
                 server.mocked_send(Message("client_ready", None, "fake_client%i" % i))
 
+            def heartbeat():
+                for i in range(5):
+                    server.mocked_send(Message("heartbeat", {'state': STATE_RUNNING, 'current_cpu_usage': 50}, "fake_client%i" % i))
+
```

Comment:
I'm trying to understand why this test broke. Was it because the workers went missing before the test ended? 

If so, maybe we could instead just decorate the test with:

```python
@mock.patch("locust.runners.HEARTBEAT_LIVENESS", new=30)
```
instead of sending the fake heartbeat messages?

Reply:
I originally thought that was why it was failing but now I'm not sure. That patch didn't work either. I'll look into this more when I get off work

Reply:
So, what's actually happening is not that fake workers go missing, instead their state is never changed from ready since they never send a heartbeat.  If all the workers are set to ready, then the state will be set to stopped on the next check.

This wasn't a problem before since there was never a check to stop after the test was started. The client_listener would just hang after it received all the fake client_ready messages. Having the heartbeat_worker check to stop was responsible for breaking the test.

However, it's no longer a problem after I implement your suggestion to move the check inside the if statement, so I'll revert the changes to this test, and write some new tests